File: integrations/crewai/nosocial_crewai/reporter.py
"""
NoSocial reporter for CrewAI — auto-reports agent interactions to the oracle.

Usage:
    from nosocial_crewai import NoSocialReporter

    reporter = NoSocialReporter(oracle_url="https://api.nosocial.me")
    crew = Crew(
        agents=[...],
        tasks=[...],
        task_callback=reporter.task_callback,
    )
"""

import hashlib
import logging
import time
import uuid
from typing import Optional

import requests

from nosocial_crewai.identity import AgentIdentity

logger = logging.getLogger("nosocial")


class NoSocialReporter:
    """Reports CrewAI task completions as NoSocial interaction reports."""

    def __init__(
        self,
        oracle_url: str = "https://api.nosocial.me",
        keys_dir: str = ".nosocial/keys",
        crew_name: str = "default-crew",
        auto_register: bool = True,
    ):
        self.oracle_url = oracle_url.rstrip("/")
        self.keys_dir = keys_dir
        self.crew_name = crew_name
        self.auto_register = auto_register
        self._identities: dict[str, AgentIdentity] = {}
        self._registered: set[str] = set()
        # The crew itself gets an identity — it's the "reporter" in most cases
        self._crew_identity = self._get_or_create_identity(f"crew:{crew_name}")

    def _get_or_create_identity(self, name: str) -> AgentIdentity:
        """Get or create a persistent identity for a named agent."""
        if name not in self._identities:
            self._identities[name] = AgentIdentity.load_or_create(
                name, self.keys_dir
            )
        return self._identities[name]

    def _ensure_registered(self, identity: AgentIdentity, name: str) -> bool:
        """Register an agent with the oracle if not already registered."""
        if identity.did in self._registered:
            return True
        if not self.auto_register:
            return False

        try:
            # Step 1: Request challenge
            resp = requests.post(
                f"{self.oracle_url}/v1/agents/challenge",
                json={"publicKey": identity.public_key_str},
                timeout=10,
            )
            if resp.status_code == 409:
                # Already registered
                self._registered.add(identity.did)
                return True
            resp.raise_for_status()
            challenge_data = resp.json()

            # Step 2: Sign challenge and register
            signature = identity.sign({"challenge": challenge_data["challenge"]})
            resp = requests.post(
                f"{self.oracle_url}/v1/agents/register",
                json={
                    "challengeId": challenge_data["challengeId"],
                    "signature": signature,
                    "publicKey": identity.public_key_str,
                    "name": name,
                },
                timeout=10,
            )
            resp.raise_for_status()
            self._registered.add(identity.did)
            logger.info(f"Registered agent '{name}' as {identity.did}")
            return True

        except Exception as e:
            logger.warning(f"Failed to register '{name}' with oracle: {e}")
            return False

    def _submit_report(
        self,
        reporter: AgentIdentity,
        subject: AgentIdentity,
        domain: str,
        score: float,
        context: Optional[dict] = None,
    ) -> bool:
        """Submit an interaction report to the oracle."""
        report = {
            "id": str(uuid.uuid4()),
            "reporter": reporter.did,
            "subject": subject.did,
            "timestamp": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
            "domain": domain,
            "score": max(-1.0, min(1.0, score)),
        }
        if context:
            report["context"] = context

        signature = reporter.sign(report)
        report["signature"] = signature

        try:
            resp = requests.post(
                f"{self.oracle_url}/v1/reports",
                json=report,
                timeout=10,
            )
            if resp.status_code == 201:
                logger.debug(
                    f"Reported: {reporter.did[:20]}... → {subject.did[:20]}... "
                    f"domain={domain} score={score}"
                )
                return True
            else:
                logger.warning(f"Oracle rejected report: {resp.json()}")
                return False
        except Exception as e:
            logger.warning(f"Failed to submit report: {e}")
            return False

    def task_callback(self, output) -> None:
        """
        CrewAI task_callback — register on a Crew to auto-report.

        Maps task completion to a NoSocial interaction report:
        - Reporter: the crew identity (the orchestrator observing the work)
        - Subject: the agent that executed the task
        - Domain: task_completion
        - Score: 1.0 for successful completion (output exists), -0.5 for empty
        - Context: task description, agent role, output length
        """
        agent_role = self._extract_agent_name(output)
        if not agent_role:
            return

        # Namespace agent identity by crew to avoid collisions
        # (e.g., "my-crew:researcher" vs "your-crew:researcher")
        agent_name = f"{self.crew_name}:{agent_role}"
        subject = self._get_or_create_identity(agent_name)

        # Ensure both parties are registered
        if not self._ensure_registered(self._crew_identity, f"crew:{self.crew_name}"):
            return
        if not self._ensure_registered(subject, agent_name):
            return

        # Determine score from output quality signals
        raw_output = getattr(output, "raw", "") or ""
        has_output = len(raw_output.strip()) > 0
        score = 0.8 if has_output else -0.5

        context = {
            "taskType": "crewai-task",
            "skillId": self._extract_task_id(output),
            "outputAccepted": has_output,
        }

        # Report task_completion
        self._submit_report(
            reporter=self._crew_identity,
            subject=subject,
            domain="task_completion",
            score=score,
            context=context,
        )

        # If the output is substantial, also report on reliability
        if has_output:
            self._submit_report(
                reporter=self._crew_identity,
                subject=subject,
                domain="reliability",
                score=0.8,
                context=context,
            )

    def _extract_agent_name(self, output) -> Optional[str]:
        """Extract agent name from TaskOutput, handling different CrewAI versions."""
        # TaskOutput.agent is typically the agent's role string
        agent = getattr(output, "agent", None)
        if agent is None:
            return None
        # In some versions agent is a string (role), in others it's an Agent object
        if isinstance(agent, str):
            return agent
        return getattr(agent, "role", None) or getattr(agent, "name", None)

    def _extract_task_id(self, output) -> str:
        """Extract a stable task identifier from TaskOutput via content hash."""
        desc = getattr(output, "description", "") or ""
        if not desc:
            return "unknown-task"
        return hashlib.sha256(desc.encode()).hexdigest()[:12]

File: integrations/ai-sdk/src/reporter.ts
/**
 * NoSocial reporter — handles oracle registration and report submission.
 *
 * Fire-and-forget by default: reports are submitted without blocking
 * the model response. Errors are passed to the onError callback.
 */

import { AgentIdentity } from "./identity.js";
import type { NoSocialOptions, ReportPayload } from "./types.js";

export class Reporter {
  private readonly oracleUrl: string;
  private readonly keysDir: string;
  private readonly agentName: string;
  private readonly autoRegister: boolean;
  private readonly onError: (error: unknown) => void;

  private identities = new Map<string, AgentIdentity>();
  private registered = new Set<string>();
  private agentIdentity: AgentIdentity | null = null;
  private initPromise: Promise<void> | null = null;

  constructor(options: NoSocialOptions = {}) {
    this.oracleUrl = (options.oracleUrl || "https://api.nosocial.me").replace(/\/$/, "");
    this.keysDir = options.keysDir || ".nosocial/keys";
    this.agentName = options.agentName || "default-agent";
    this.autoRegister = options.autoRegister ?? true;
    this.onError = options.onError || (() => {});
  }

  private async init(): Promise<void> {
    if (!this.initPromise) {
      this.initPromise = (async () => {
        this.agentIdentity = await this.getOrCreateIdentity(`agent:${this.agentName}`);
      })();
    }
    return this.initPromise;
  }

  private async getOrCreateIdentity(name: string): Promise<AgentIdentity> {
    let identity = this.identities.get(name);
    if (!identity) {
      identity = await AgentIdentity.loadOrCreate(name, this.keysDir);
      this.identities.set(name, identity);
    }
    return identity;
  }

  private async ensureRegistered(identity: AgentIdentity, name: string): Promise<boolean> {
    if (this.registered.has(identity.did)) return true;
    if (!this.autoRegister) return false;

    try {
      // Step 1: Request challenge
      const challengeResp = await fetch(`${this.oracleUrl}/v1/agents/challenge`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ publicKey: identity.publicKeyStr }),
        signal: AbortSignal.timeout(10_000),
      });

      if (challengeResp.status === 409) {
        const data = await challengeResp.json() as { error?: string };
        const msg = (data.error || "").toLowerCase();
        if (msg.includes("already registered") || msg.includes("already exists")) {
          this.registered.add(identity.did);
          return true;
        }
        return false;
      }
      if (!challengeResp.ok) return false;

      const challengeData = await challengeResp.json() as {
        challengeId: string;
        challenge: string;
      };

      // Step 2: Sign challenge and register
      const signature = identity.sign({ challenge: challengeData.challenge });
      const registerResp = await fetch(`${this.oracleUrl}/v1/agents/register`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          challengeId: challengeData.challengeId,
          signature,
          publicKey: identity.publicKeyStr,
          name,
        }),
        signal: AbortSignal.timeout(10_000),
      });

      if (registerResp.status === 409 || registerResp.ok) {
        this.registered.add(identity.did);
        return true;
      }
      return false;
    } catch {
      return false;
    }
  }

  private async submitReport(
    reporter: AgentIdentity,
    subject: AgentIdentity,
    domain: ReportPayload["domain"],
    score: number,
    context?: Record<string, unknown>,
  ): Promise<void> {
    const report: Record<string, unknown> = {
      id: crypto.randomUUID(),
      reporter: reporter.did,
      subject: subject.did,
      timestamp: new Date().toISOString(),
      domain,
      score: Math.max(-1, Math.min(1, score)),
    };
    if (context) report.context = context;

    report.signature = reporter.sign(report as Record<string, unknown>);

    const resp = await fetch(`${this.oracleUrl}/v1/reports`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(report),
      signal: AbortSignal.timeout(10_000),
    });

    if (!resp.ok) {
      throw new Error(`Oracle rejected report: ${resp.status}`);
    }
  }

  /**
   * Report an event. Fire-and-forget: errors go to onError callback.
   */
  reportEvent(
    subjectName: string,
    domain: ReportPayload["domain"],
    score: number,
    context?: Record<string, unknown>,
  ): void {
    // Fire and forget — don't block the caller
    this._doReport(subjectName, domain, score, context).catch(this.onError);
  }

  private async _doReport(
    subjectName: string,
    domain: ReportPayload["domain"],
    score: number,
    context?: Record<string, unknown>,
  ): Promise<void> {
    await this.init();
    const reporter = this.agentIdentity!;
    const subject = await this.getOrCreateIdentity(`${this.agentName}:${subjectName}`);

    if (!await this.ensureRegistered(reporter, `agent:${this.agentName}`)) return;
    if (!await this.ensureRegistered(subject, `${this.agentName}:${subjectName}`)) return;

    await this.submitReport(reporter, subject, domain, score, context);
  }
}
